components/script/dom/notification.rs

```diff
@@ -0,0 +1,545 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use dom_struct::dom_struct;
+use js::jsapi::Heap;
+use js::jsval::JSVal;
+use js::rust::{HandleObject, MutableHandleValue};
+use servo_url::{ImmutableOrigin, ServoUrl};
+
+use super::permissionstatus::PermissionStatus;
+use crate::dom::bindings::callback::ExceptionHandling;
+use crate::dom::bindings::codegen::Bindings::NotificationBinding::{
+    NotificationAction, NotificationDirection, NotificationMethods, NotificationOptions,
+    NotificationPermission, NotificationPermissionCallback,
+};
+use crate::dom::bindings::codegen::Bindings::PermissionStatusBinding::PermissionStatus_Binding::PermissionStatusMethods;
+use crate::dom::bindings::codegen::Bindings::PermissionStatusBinding::{
+    PermissionDescriptor, PermissionName, PermissionState,
+};
+use crate::dom::bindings::codegen::UnionTypes::UnsignedLongOrUnsignedLongSequence;
+use crate::dom::bindings::error::Error;
+use crate::dom::bindings::import::module::{Fallible, Rc};
+use crate::dom::bindings::inheritance::Castable;
+use crate::dom::bindings::reflector::reflect_dom_object_with_proto;
+use crate::dom::bindings::root::{Dom, DomRoot};
+use crate::dom::bindings::str::{DOMString, USVString};
+use crate::dom::bindings::trace::RootedTraceableBox;
+use crate::dom::bindings::utils::to_frozen_array;
+use crate::dom::eventtarget::EventTarget;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::permissions::{get_descriptor_permission_state, PermissionAlgorithm, Permissions};
+use crate::dom::promise::Promise;
+use crate::dom::serviceworkerglobalscope::ServiceWorkerGlobalScope;
+use crate::dom::serviceworkerregistration::ServiceWorkerRegistration;
+use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
+
+// TODO: Service Worker API (persistent notification)
+// https://notifications.spec.whatwg.org/#service-worker-api
+
+/// <https://notifications.spec.whatwg.org/#notifications>
+#[dom_struct]
+pub(crate) struct Notification {
+    eventtarget: EventTarget,
+    /// <https://notifications.spec.whatwg.org/#service-worker-registration>
+    serviceworker_registration: Option<Dom<ServiceWorkerRegistration>>,
+    /// <https://notifications.spec.whatwg.org/#concept-title>
+    title: DOMString,
+    /// <https://notifications.spec.whatwg.org/#body>
+    body: DOMString,
+    /// <https://notifications.spec.whatwg.org/#data>
+    #[ignore_malloc_size_of = "mozjs"]
+    data: Box<Heap<JSVal>>,
```

Comment:
Storing a `Heap` directly inside of DOM objects like this is both safe and avoids an extra box allocation.